src/expr/core/src/error.rs

```diff
@@ -117,6 +117,37 @@ pub enum ExprError {
 
     #[error("invalid state: {0}")]
     InvalidState(String),
+
+    #[error("error in cryptography: {0}")]
+    Cryptography(Box<CryptographyError>),
+}
+
+#[derive(Debug)]
+pub enum CryptographyStage {
+    Encrypt,
+    Decrypt,
+}
+
+#[derive(Debug)]
+pub struct CryptographyError {
+    pub stage: CryptographyStage,
+    pub payload: Box<[u8]>,
+    pub reason: Box<str>,
+}
```

Comment:
```suggestion
#[derive(Debug, thiserror::Error)]
#[error("{stage:?} stage, reason: {reason:?}")]
pub struct CryptographyError {
    pub stage: CryptographyStage,
    #[source]
    pub reason: ErrorStack,
}
```

* As the dylint rule suggested, embed source error rather than its string form.
* It may not be proper to include plaintext or ciphertext in logs, although it could be helpful for troubleshooting. Even if we do want to include it, a simple `{payload:02X?}` is enough. No need to fold to a String manually.